src/v/datalake/event_multiplexer.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "container/fragmented_vector.h"
+#include "datalake/schemaless_translator.h"
+#include "json/writer.h"
+#include "model/record.h"
+
+// TODO Only needed for method definitions, not class. Move methods out of line
+// if possible
+#include "storage/parser_utils.h"
+
+#include <seastar/core/future.hh>
+
+#include <optional>
+
+namespace datalake {
+/*
+Consumes logs and sends events to the appropriate translator
+based on the schema ID. This is meant to be called with a
+read_committed_reader created from a kafka::partition_proxy.
+
+This is a skeleton class. Currently it only uses the trivial
+schemaless_translator.
+
+pseudocode:
+for each event {
+    t = get_translator(event.schema_id);
+    d = t.translate(event);
+    w = get_data_writer(event.schema_id);
+    w.add_event(d);
+}
+
+The WriterFactory is a class with
+- instance method `create_writer`
+- nested class `data_writer_result`
+- nested class `data_writer`
+  - instance method `add_data_struct` to consume a parsed event
+  - instance method `finish` to clean up and return a `data_writer_result`
+*/
+template<typename WriterFactory>
+class event_multiplexer {
+public:
+    event_multiplexer(WriterFactory& writer_factory)
+      : _translator{schemaless_translator()}
+      , _writer_factory{writer_factory} {}
+    ss::future<ss::stop_iteration> operator()(model::record_batch&& batch);
+    ss::future<fragmented_vector<typename WriterFactory::data_writer_result>>
+    end_of_stream();
+
+private:
+    using translator = std::variant<schemaless_translator>;
+
+    translator& get_translator() { return _translator; }
+    WriterFactory::data_writer& get_writer() {
+        auto schema = std::visit(
+          [](schemaless_translator& tr) { return tr.get_schema(); },
+          _translator);
+        if (!_writer) {
+            _writer = _writer_factory.create_writer(std::move(schema));
+        }
```

Comment:
We can return early if there's an existing writer without having to construct a schema

Reply:
Done

Reply:
Done